core-bundle/src/Resources/contao/classes/DataContainer.php

```diff
@@ -858,6 +860,29 @@ protected function switchToEdit($id)
 		return $strUrl . (!empty($arrKeys) ? '&' : '') . (Input::get('table') ? 'table=' . Input::get('table') . '&amp;' : '') . 'act=edit&amp;id=' . rawurlencode($id);
 	}
 
+	/**
+	 * @throws AccessDeniedException
+	 */
+	protected function denyAccessUnlessGranted($attributes, $subject): void
+	{
+		/** @var Security $security */
+		$security = System::getContainer()->get('security.helper');
+
+		if ($security->isGranted($attributes, $subject))
+		{
+			return;
+		}
+
+		$message = 'Access denied.';
+
+		if ($subject instanceof DataContainerSubject)
+		{
+			$message = sprintf('Access denied to %s [%s].', $subject, $attributes);
+		}
+
+		throw new AccessDeniedException($message);
```

Comment:
subject and attributes should be added to the exception (according to `Controller::denyAccessUnlessGranted()`

```
$exception->setAttributes($attribute);
$exception->setSubject($subject);
```

Reply:
Done in 0a9dffbbc7c237bdbf09a8bc3f2bf22ff961462a.